nvflare/private/fed/server/job_validator.py

```diff
@@ -0,0 +1,150 @@
+
+        # Default or empty study is always valid
+        if not study_name or study_name == Study.DEFAULT_STUDY_NAME:
+            return
+
+        study = self.study_manager.get_study(study_name, self.fl_ctx)
+        if not study:
+            raise ValueError(f"Study {study_name} does not exist")
+
+        # Validating all clients are participants of the study
+        all_clients = self._get_all_clients(meta)
+        if study.participating_clients:
+            participants = set(study.participating_clients)
+            if not all_clients.issubset(participants):
+                diff = all_clients - participants
+                raise ValueError("Following clients are not participants of the study {}: {}".format(study_name, diff))
+
+        # Validating study validity period
+        current_time = time.time()
+        if study.start_time and study.start_time > current_time:
+            raise ValueError(
+                f"Study {study_name} has not started yet. "
+                f"The start time is {datetime.utcfromtimestamp(study.start_time).isoformat()}"
+            )
+
+        if study.end_time and study.end_time < current_time:
+            raise ValueError(
+                f"Study {study_name} has ended. "
+                f"The end time is {datetime.utcfromtimestamp(study.end_time).isoformat()}"
+            )
+
+    def _validate_deployment(self, meta: dict, zip_file: ZipFile) -> None:
+
+        job_name = meta.get(JobMetaKey.JOB_FOLDER_NAME)
+        deploy_map = meta.get(JobMetaKey.DEPLOY_MAP)
+        if not deploy_map:
+            raise ValueError(f"deploy_map is missing in meta.json for job {job_name}")
+
+        # Validating mandatory clients are deployed
+        mandatory_clients = meta.get(JobMetaKey.MANDATORY_CLIENTS)
+        if mandatory_clients:
+            mandatory_set = set(mandatory_clients)
+            all_clients = self._get_all_clients(meta)
+            if all_clients and not mandatory_set.issubset(all_clients):
+                diff = mandatory_set - all_clients
+                raise ValueError(f"Mandatory clients {diff} are not in the deployment for job {job_name}")
+
+        # Validating all apps exists
+        client_list = []
+        for app, deployment in deploy_map.items():
+            zip_folder = job_name + "/" + app + "/config/"
+            try:
+                zip_file.getinfo(zip_folder)
+            except KeyError:
+                raise ValueError(f"App {app} in deploy_map doesn't exist for job {job_name}")
+            client_list.extend([site for site in deployment if site != "server"])
+
+        # Make sure an app is not deployed to multiple clients
+        if client_list:
+            duplicates = [site for site, count in collections.Counter(client_list).items() if count > 1]
+            if duplicates:
+                raise ValueError(f"More than one app are running on clients {duplicates} for job {job_name}")
```

Comment:
Suggest change the error message since no job/app is running at the moment that it is being checked.

Reply:
Changed.